keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,606 @@ yet but we think may be interesting to consider in the future.
+concurrently.  The mutating and readonly requests are effectively
+commingled, with the readonly requests having width 1 and the mutating
+requests having width 2.  The primary resource limit applied is that
+at any moment in time the weighted sum `1 * (number of running
+non-zero priority readonly requests) + 2 * (number of running non-zero
+priority mutating requests)` should not exceed the concurrency limit.
+Requests of priority zero are neither counted nor limited, as in
+today's max-in-flight handler.
+
+At the first stage of development, an apiserver’s commingled
+concurrency limit will be derived from the existing configuration
+options for max-mutating-in-flight and max-readonly-in-flight, by
+taking their weighted sum.  Later we may migrate to a single direct
+configuration option.
+
+#### Secondary Memory Protection
+
+A RequestPriority is also configured with a limit on the number of
+requests that may be waiting in a given queue.
+
+#### Latency Protection
+
+An apiserver is also configured with a limit on the amount of time
+that a request may wait in its queue.  If this time passes while a
+request is still waiting for service then the request will be
+rejected.
+
+### Queuing
+
+Once a request is categorized and assigned to a queue the next
+decision is whether to reject or accept that request.
+
+A request of priority zero is never rejected and never waits in a
+queue; such a request is dispatched as soon as it arrives.
+
+For queuing requests of non-zero priority, the first step is to reject
+all the requests that have been waiting longer than the configured
+limit.  Once that is done, the newly arrived request is considered.
+This request is rejected if and only if the total number of requests
+waiting in its queue is at least the configured limit on that number.
+
+A possible alternative would accept the request unconditionally and,
+if that made the queue too long, reject the request at the head of the
+queue.  That would be the preferred design if we were confident that
+rejection will cause the client to slow down.  Lacking that
+confidence, we choose to reject the youngest rather than the oldest
+request of the queue, so that an investment in holding a request in a
+queue has a chance of eventually getting useful work done.
+
+### Dispatching
+
+Requests of priority zero are never held up in a queue; they are
+always dispatched immediately.  Following is how the other requests
+are dispatched at a given apiserver.
+
+Whenever the width-weighted sum of running non-zero priority requests
+is below the concurrency limit, it is time to consider dispatching
+another request for service.  This is done in two stages.  First, a
+priority level with non-empty queues is chosen. The logic for doing
+that makes use of each priority level's _assured concurrency value_
+(ACV), which is calculated as follows.
+
+```
+ACV(l) = ceil( SCL * ACS(l) / (100 + sum[priority levels k] ACS(k) ) )
```

Comment:
why is it 100? i have a complex feeling about introducing more magic numbers..

Reply:
This number is not particularly magic.  It just establishes a magnitude standard to compare the ACS values with.  All that matters is their ratios.  We _could_ create another configuration parameter to define the share for shared concurrency.  But, without loss of generality, we can fix it at a particular number.